### ISSUE TITLE
Erroneous results from main_fancy123_refine.py

### ISSUE THREAD

**Issue by ssinha89** [YuQiao0303/Fancy123#4]
@YuQiao0303 I have been able to run fix the problems in https://github.com/YuQiao0303/Fancy123/issues/2 by replacing the meshes folder output of run_init.py with the outputs from InstantMesh code repository. However running  main_fancy123_refine.py leads to the following outputs, with incorrect geometry and texture refinement (output files attached below). It seems the texture unprojection is problematic (eyes of the unicorn is mapped elsewhere), although the InstantMesh output is fine. The final_mesh.obj is tiny in comparison to the initial Instantmesh.obj Request you to kindly assist in solving this issue.
The following are the output files

**1. Instantmesh.obj**

<img width="2796" height="2965" alt="Image" src="https://github.com/user-attachments/assets/d12b825c-180e-4e9c-96c8-b0ad7e4edab3" />
-----

**2. unproj_wo_2D_deform.obj**

![Image](https://github.com/user-attachments/assets/8fc91704-cef6-4819-b0f9-6a3ffc059611)
-----
**3. 2D_deform_deformed.obj**

![Image](https://github.com/user-attachments/assets/4df45195-e4db-4405-a9fd-9c951d894389)
-----
**4. 3D_deformed_mesh.obj / final_mesh.obj**

![Image](https://github.com/user-attachments/assets/72f2897a-3b10-4325-a7f5-5bd82a3418e2)
-----
**5.  2D_deform_deformed_images.png**

![Image](https://github.com/user-attachments/assets/235a2686-7335-449c-aa57-bccdf6a1300e)
-----
**6.  3D_mesh_deform.png**

![Image](https://github.com/user-attachments/assets/b2917667-403b-41a8-9311-0bf117de41c3)

**Comment by YuQiao0303** [comment]
Hi! 

- Fancy123's init mesh is a little different from the original InstantMesh, e.g. we subdivide the mesh for higher resolution. However this looks like some other issue. 
- Is it possible for you to upload your whole folder to GitHub (e.g. namely 'Fancy123_debug'), including all the scripts, the replaced instantmesh.obj file, and all your output files? That would enable me to download and try to reproduce this issue for easier troubleshooting.
- Also, do you use Docker? I'm considering providing a Docker image for Fancy123.

**Comment by ssinha89** [comment]
@YuQiao0303 I am using exactly the same script main_fancy_123_refine.py in this git repo, after using the default git run.py of InstantMesh to produce the output files on the sample data. I did not make any changes to the code or the config files in either repository. I simply replaced the meshes folder generated by run_init.py with the meshes folder from InstantMesh because run_init.py wasn't working.

**Comment by ssinha89** [comment]
@YuQiao0303 Setting debug=True in run_mesh_refine() in fancy123/optimize/mesh_optimize.py produces the following output in output file optimize_mesh_normal.gif

![Image](https://github.com/user-attachments/assets/fa3977bd-5ecd-4e1a-abf0-210e8ebc353f)

optimize_mesh_normal.gif is written via Lines 407-414 in run_mesh_refine(). 

            merge_normal =(target_images + images)/2
            cat_normal = torch.cat([images,target_images],dim=1)
            cat_normal = torch.cat([cat_normal,merge_normal],dim=1)
            # now cat images of different cameras
            cat_normal = torch.cat(torch.unbind(cat_normal, dim=0), dim=1)
            cat_normal = cat_normal.clip(0,1).detach().squeeze().cpu().numpy()
            cat_normal = img_as_ubyte(cat_normal)[...,:3]
            normal_writer.append_data(cat_normal)

From the output it seems that the rendered variable image is in a pose different from the target normal variable target_images. Please suggest what might be causing the issue. Is the variable camera_positions_np being correctly computed from the following azimuth and elevations in configs/instantmesh.yaml ?

azim_list: [30.0, 90.0, 150.0, 210.0, 270.0, 330.0]
elevations: [20.0, -10.0, 20.0, -10.0, 20.0, -10.0]

**Comment by YuQiao0303** [comment]
> [@YuQiao0303](https://github.com/YuQiao0303) Setting debug=True in run_mesh_refine() in fancy123/optimize/mesh_optimize.py produces the following output in output file optimize_mesh_normal.gif
> 
> ![Image](https://github.com/user-attachments/assets/fa3977bd-5ecd-4e1a-abf0-210e8ebc353f)
> 
> optimize_mesh_normal.gif is written via Lines 407-414 in run_mesh_refine().
> 
> ```
>         merge_normal =(target_images + images)/2
>         cat_normal = torch.cat([images,target_images],dim=1)
>         cat_normal = torch.cat([cat_normal,merge_normal],dim=1)
>         # now cat images of different cameras
>         cat_normal = torch.cat(torch.unbind(cat_normal, dim=0), dim=1)
>         cat_normal = cat_normal.clip(0,1).detach().squeeze().cpu().numpy()
>         cat_normal = img_as_ubyte(cat_normal)[...,:3]
>         normal_writer.append_data(cat_normal)
> ```
> 
> From the output it seems that the rendered variable image is in a pose different from the target normal variable target_images. Please suggest what might be causing the issue. Is the variable camera_positions_np being correctly computed from the following azimuth and elevations in configs/instantmesh.yaml ?
> 
> azim_list: [30.0, 90.0, 150.0, 210.0, 270.0, 330.0] elevations: [20.0, -10.0, 20.0, -10.0, 20.0, -10.0]

Hi! By comparing the code of Fancy123 and InstantMesh, I find the reason: In Fancy123, we modify the coordinate system of InstantMesh by the code in the figure below:

<img width="1878" height="784" alt="Image" src="https://github.com/user-attachments/assets/3d273df1-7646-4dfb-9e4d-051142878b7b" />


In order to use InstantMesh's generated mesh in Fancy123, you may need to:
1. add the above transformation code as shown in the figure before saving InstantMesh's mesh
2. Also, you may notice that we add a subdivision for InstantMesh's initial mesh. You can compare the code of src/models/lrm_mesh.py in both Fancy123 and InstantMesh, and add the corresponding subdivision code in your InstantMesh's code.
3. Since the multiview diffusion can generate different results each time, (we reset random seeds in our Fancy123 but InstantMesh only resets random seeds once at the beginning), so you may also need to put InstantMesh's generated multiview images into Fancy123's corresponding folder, then use Fancy123 to further generate normal maps and conduct background removal. Otherwise, if you only replace Fancy123's initial mesh by InstantMesh's, but keeps Fancy123's multiview images, the mesh and the images may be inconsistent.


I'm busy recently, but I will try to push a new version of Fancy123 to support the above stuff. I'm also working on providing a Docker image to address the gray box issue, so that you don't have to use InstantMesh's output and copy them here.

It is likely to take a while for me to complete the above stuff. If you're in a hurry, you can try the above instructions to try to solve the current issue.

Thank you!

**Comment by YuQiao0303** [comment]
Hi! I've tried a new server with Torch 2.8.0+cu128, Python  3.12, but I still cannot reproduce this issue.
Here's how I prepared the environment:

```bash
pip install ninja
pip install torch==2.8.0 torchvision==0.23.0 torchaudio==2.8.0 xformers==0.0.32.post2 --index-url https://download.pytorch.org/whl/cu128  # cu128/xformers-0.0.32.post2-cp39-abi3-manylinux_2_28_x86_64.whl (117.2 MB) 


pip install kaolin -f  https://nvidia-kaolin.s3.us-east-2.amazonaws.com/torch-2.8.0_cu128/kaolin-0.18.0-cp312-cp312-linux_x86_64.whl
pip install requirements.txt

pip install https://data.pyg.org/whl/torch-2.8.0%2Bcu128/torch_cluster-1.6.3%2Bpt28cu128-cp312-cp312-linux_x86_64.whl

pip install https://data.pyg.org/whl/torch-2.8.0%2Bcu128/torch_scatter-2.1.2%2Bpt28cu128-cp312-cp312-linux_x86_64.whl

pip install https://data.pyg.org/whl/torch-2.8.0%2Bcu128/torch_sparse-0.6.18%2Bpt28cu128-cp312-cp312-linux_x86_64.whl

```
